### ISSUE TITLE
Kubectl proxy would strip out the parameter for websocket request

### ISSUE THREAD

**Issue by 1wpro2** [kubernetes/kubernetes#89360]
**What happened**:
Create a deployment to run a websocket server. Use `kubectl proxy` for accessing the service outside of the cluster.

**What you expected to happen**:

Invoke the URL, for example as below, would pas in the token to the websocket server.
```
"ws://..../api/v1/namespaces/..../services/theservice/proxy/websocket?token=123"
```

However, instead, the parameter is stripped out.

![image](https://user-images.githubusercontent.com/52658745/77302773-ce93c200-6d2c-11ea-9d94-c1f085ef5215.png)

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.17.1
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Darwin
- Kernel (e.g. `uname -a`): Darwin Kernel Version 19.2.0:
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by 1wpro2** [comment]
looks like the same issue with:
https://github.com/kubernetes/kubernetes/issues/77142
https://github.com/kubernetes/kubernetes/issues/67878

**Comment by 1wpro2** [comment]
actually, it looks like it's the same for apiserver proxy. for example, if accessing the apiserver directly:

```
wss://$APISERVER/api/v1/namespaces/..../services/theservice/proxy/websocket?token=123
```

![image](https://user-images.githubusercontent.com/52658745/77392510-16baef00-6dd6-11ea-8f75-f72460231aa8.png)

**Comment by brianpursley** [comment]
I thought I found a way to fix this, but I was wrong :disappointed: 

/unassign

**Comment by brianpursley** [comment]
For anyone else who works on this, here is what I tried:

Link to the relevant place in the code where the querystring is lost:
https://github.com/brianpursley/kubernetes/blob/25ebb388f033d17a35c8bdec8b8224406e856215/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L263

My attempted change:
```
        location := *h.Location
	if h.UseRequestLocation {
		location = *req.URL
		location.Scheme = h.Location.Scheme
		location.Host = h.Location.Host
	} else {  // <-- I ADDED THIS ELSE CONDITION
		// Even when not using the request location, we want to preserve the querystring
		location.RawQuery = req.URL.RawQuery
	}
```

This worked when I tested it locally and my unit test passed, but it broke A LOT of things in e2e testing so clearly this is used by more than just kubectl proxy.

**Comment by 1wpro2** [comment]
> For anyone else who works on this, here is what I tried:
> 
> Link to the relevant place in the code where I think the querystring is lost:
> https://github.com/kubernetes/kubernetes/blob/cf1b72e4571b69995ce65a659a9b67c867b1d1b0/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L263-L268
> 
> My attempted change:
> 
> ```
>         location := *h.Location
> 	if h.UseRequestLocation {
> 		location = *req.URL
> 		location.Scheme = h.Location.Scheme
> 		location.Host = h.Location.Host
> 	} else {  // <-- I ADDED THIS ELSE CONDITION
> 		// Even when not using the request location, we want to preserve the querystring
> 		location.RawQuery = req.URL.RawQuery
> 	}
> ```
> 
> This worked when I tested it locally and my unit test passed, but it broke A LOT of things in e2e testing so clearly this is used by more than just kubectl proxy.

@brianpursley, I had looked in the same file, however, I think this is still within `kubectl proxy`, while the ultimate issue lies in apiserver.

Have you been able to run your own apiserver, to replace the inbuilt from minikube? That was where the last time I have stopped if I remember correctly.

**Comment by brianpursley** [comment]
@1wpro2 yeah, I'm using kind, but running using hack/local-up-cluster.sh and running a custom apiserver.

the apimachinery code is used on both client and server I believe.

I got some more time and looked into it last night and think I was actually on the right track before.  I made my change again and all e2e's are passing, but not completely satisfied with how I made the change yet so I'm still working on it.  Hope to have something soon.

/assign
